avm/src/vm/opcode.rs

```diff
@@ -222,2012 +224,2014 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
+    Ok(None)
   });
   cpu!(gtestr => fn(args, hand_mem) {
-    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]));
-    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1]));
+    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0])?)?;
+    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1])?)?;
     let out = if a_str >= b_str { 1i64 } else { 0i64 };
-    hand_mem.write_fixed(args[2], out);
-    None
+    hand_mem.write_fixed(args[2], out)?;
+    Ok(None)
   });
 
   // String opcodes
   cpu!(catstr => fn(args, hand_mem) {
-    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]));
-    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1]));
+    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0])?)?;
+    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1])?)?;
     let out_str = format!("{}{}", a_str, b_str);
-    hand_mem.write_fractal(args[2], &HandlerMemory::str_to_fractal(&out_str));
-    None
+    hand_mem.write_fractal(args[2], &HandlerMemory::str_to_fractal(&out_str))?;
+    Ok(None)
   });
   cpu!(split => fn(args, hand_mem) {
-    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]));
-    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1]));
+    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0])?)?;
+    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1])?)?;
     let out_hms = a_str.split(&b_str).map(|out_str| HandlerMemory::str_to_fractal(&out_str));
-    hand_mem.init_fractal(args[2]);
+    hand_mem.init_fractal(args[2])?;
     for out in out_hms {
-      hand_mem.push_fractal(args[2], out);
+      hand_mem.push_fractal(args[2], out)?;
     }
-    None
+    Ok(None)
   });
   cpu!(repstr => fn(args, hand_mem) {
-    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]));
-    let n = hand_mem.read_fixed(args[1]);
+    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0])?)?;
+    let n = hand_mem.read_fixed(args[1])?;
     let out_str = a_str.repeat(n as usize);
-    hand_mem.write_fractal(args[2], &HandlerMemory::str_to_fractal(&out_str));
-    None
+    hand_mem.write_fractal(args[2], &HandlerMemory::str_to_fractal(&out_str))?;
+    Ok(None)
   });
   cpu!(matches => fn(args, hand_mem) {
-    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0]));
-    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1]));
-    let b_regex = Regex::new(&b_str).unwrap();
+    let a_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0])?)?;
+    let b_str = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1])?)?;
+    let b_regex = Regex::new(&b_str).map_err(|regex_err| VMError::Other(format!("Bad regex construction: {}", regex_err)))?;
```

Comment:
This is good, but makes me wonder what we can/should do at compile time for this. A bad regex should be detectable during the compilation phase.

Reply:
unfortunately not :/ regex does some heap allocations that can't be done at compile time* - see <https://docs.rs/regex/1.4.5/src/regex/exec.rs.html>

\* allocations in const-fns is planned, per <https://github.com/rust-lang/const-eval/issues/20>

Reply:
Oh, I was thinking something like "Should we have a fake rust file that we inject the users' regex into and test if it parses correctly and just emit an error during compilation when the regex string is static"

But maybe the correct answer here is: "the regex API in Alan needs to be changed to be a `Result` and we return an Error to the user if it fails to build correctly"

Reply:
ah I see - we could also ship `ripgrep` with our compiler binary, and when we detect regex usage we
1. create an empty file
2. run `rg` on that empty file using the user's regex
3. continue compiling if the exit code is 0 or throw a compiler error if the exit code is non-zero

We could also shave `regex` off our dependency tree and save some space in the binary size by just shipping `ripgrep` in the avm regardless of if the compiler is there, and every time the user wants to do a regex query we do the same but on a file containing the input.

Reply:
In all seriousness though, we *might* be able to pre-compile the Regex in the compiler by calling `regex` as a wasm module, store the object as a `Buffer`, and include it in the aga/agc output, but that would take a solid amount of time to get working...

Reply:
actually it's not worth it to store the memory representation of `regex::Regex`, but we could still run it from wasm to validate the input

Reply:
I think all of this makes it clear that we need to change the regex-related APIs in Alan to be `Result`-based so it can provide the failure at runtime. At least the user will need to have defined some code to handle what happens if it can't parse the regex, though that is almost certainly going to be a logic bug. :/